unstable/wlr-workspace-unstable-v1.xml

```diff
@@ -0,0 +1,192 @@
+        This event indicates that the compositor is done sending events to the
+        zwlr_workspace_manager_v1. The server will destroy the object immediately after
+        sending this request, so it will become invalid and the client should free any
+        resources associated with it.
+      </description> 
+    </event> 
+
+    <request name="stop">
+      <description summary="stop sending events">
+        Indicates the client no longer wishes to receive events for new workspaces.
+        However the compositor may emit further workspace events, until the finished
+        event is emitted.
+
+        The client must not send any more requests after this one.
+      </description>
+    </request>
+  </interface>
+
+  <interface name="zwlr_workspace_handle_v1" version="1">
+    <description summary="a workspace handing a group of surfaces">
+      A zwlr_workspace_handle_v1 object represents a a workspace that handles a group of
+      surfaces.
+
+      Each workspace has a name, conveyed to the client with the name event; a list of
+      states, conveyed to the client with the state event; a list of outputs, conveyed
+      via one or more output events; and optionally a set of coordinates, conveyed to
+      the client with the coordinates event. The client may request that the compositor
+      activate or deactivate the workspace.
+    </description>
+
+    <event name="output">
+      <description summary="output added to workspace">
+        This event is emitted whenever an output is added to the workspace.
+      </description>
+      <arg name="output" type="object" interface="wl_output"/>
+    </event>
+
+    <event name="output_remove">
+      <description summary="output removed from workspace">
+        This event is emitted whenever an output is removed from the workspace.
+      </description>
+      <arg name="output" type="object" interface="wl_output"/>
+    </event>
+
+    <event name="name">
+      <description summary="workspace name changed">
+        This event is emitted immediately after the zwlr_workspace_handle_v1 is created
+        and whenever the name of the workspace changes.
+      </description>
+      <arg name="number" type="int"/>
+    </event>
+
+    <event name="coordinates">
+      <description summary="workspace coordinates changed">
+        If the compositor assigns coordinates to workspaces, this event is emitted
+        immediately after the zwlr_workspace_handle_v1 is created and whenever the
+        coordinates of the workspace change. Coordinates have an arbitrary number of
+        dimensions with an unsigned integer position for each dimension. Among a set of
+        workspaces that have identical sets of outputs, workspaces must have unique
+        coordinates with the same number of dimensions, such that the workspaces form a
+        grid with at most one workspace in each position. By convention, 2D coordinates
+        represent [y_pos, x_pos].
```

Comment:
That's unconventional, is there a reason to do this?

Reply:
Row before column is the convention for image pixel coordinates and matrix indexing as far as I'm aware. I might change the text to say [row, column] instead though, since x and y are more often used for continuous variables where the convention is opposite.

Reply:
I'd prefer `x, y, z, …` because that way `pos[0]` is always x, `pos[1]` is always y and so on.

I'm still not sure it's a good idea to have this in the protocol.

Reply:
I can just leave it out, just there's a risk that different compositors will choose different conventions, and then clients won't be able to represent them graphically. Happy to go with [column, row] too.

Reply:
What if someone wants to make 3D workspaces, or even 4D? I'd suggest adding an event which has `dimension` and `extent`.

Reply:
wl_array already contains the size of the array

Reply:
Derp. I don't like wl_array, I'd rather raise the event several times.

Reply:
No need to make things more complicated because you personally don't like wl_array.

Reply:
What I suggest is more type-safe and doesn't involve manually reasoning about the type of an opaque array.

Reply:
About the index ordering, the convention I'm used to is that you would consider a 3D array to be an array of 2D arrays, such that the new index going from 2D to 3D gets inserted at the start. So it would be [z, y, x]. 

How about:
```For the purposes of representing grids of workspaces graphically, for dimension > 1 clients may interpret the final two indices as [row_index, column_index] of the workspace in a two-dimensional grid.```

I'm happy with switching this to "first two indices represent [column, row]" but I do think that is the less common convention. And I'm leaning toward saying the convention should be mentioned in the protocol, otherwise clients will not be able to reliably represent the grids in a way that will match how the compositor is representing them.

Reply:
I think it should be ordered x, y, z, w, ...; rather than with respect to the implementation details.

Reply:
an extent event at least would be necessary if the grid can have empty slots - you cant infer the extent from existing workspaces in any case if there are allowed to be gaps.

The dimensionality change events would simplify client code at the expense of more events in the protocol. I've been leaning away from this (for example removing the `empty` state since this can be inferred from other information), but is there is a precedent in wayland as to whether the protocols should provide redundant information for the sake of simplifying client code? I would have guessed that there is an overall attitude regarding this one way or the other. (and I would have guessed it was *not* to provide redundant info even if it simplifies things for the client).

Reply:
I'm not suggesting adding more events - just a single event with two parameters, dimension and extent, which is raised multiple times in a single atomic transaction.

I'm not sure there's any precedent for protocols attempting to solve a problem as generic as "workspaces".

Reply:
Alright, I'l put an end to the bikeshedding and declare it's [x, y] in the first two indices.

Reply:
That fails to address the concerns about compositors with different dimensionalities.

Reply:
What do you mean it's raised multiple times? Wouldn't it only be raised when the dimension or extent change? I must be missing something.

Reply:
Compositors with higher dimensionalities may use more indices, but the client can interpret the first two as [x, y]. So for a 3D array, the client can interpret the space as a stack of 2D grids, with the first two indices being [x, y] and the third index being which grid is displayed at any time (call this z if you like).

But I could change the language to be more generic: "if representing graphically, clients may interpret coords as x, y, z, etc in a spatial grid" or some such, rather than only mentioning [x, y] two.

Reply:
But where does the third index come from in the protocol? Also, I prefer this language:

>By convention this event is used to organize workspaces spatially into an N-dimensional grid. By convention, the first dimension is X, the second Y, the third Z, and so on. The compositor may chose to utilize these events for a more novel workspace layout convention, however. No guarantee is made about the grid being filled or bounded; there may be a workspace at coordinate 1 and another at coordinate 1000 and none in between.

Reply:
And compositors with lower dimensionality don't need a convention - if they want to specify that their 1D list of workspaces should be considered horizontally or vertically arranged, they can use 2D coords with a single row or a single column. If they don't care and it's just "next/previous" without a geometric interpretation then they can just give a 1D index.

Reply:
Forcing 1D compositors to contort themselves into two dimensions is a bad protocol. That doesn't consider 3D compositors, either.

Reply:
I like your proposed language, but I don't see how we're disagreeing, i thought that was what I was saying already. I'll just adopt your language and then I suppose we'll be happy on that point.

Reply:
If a compositor cares about whether it's workspaces are represented horizontally or vertically, then it is 2D in a sense, even if there is only one row or column. Leaving this info in the coordinates seems like a better protocol than having separate events telling you how to interpret 1D coordinates geometrically. But this comes again to whether it's a good idea to have redundancy for the sake of convenience.

Reply:
Oh I see what you were saying. By only specifying the convention for 2D and higher, the meaning of a 1D coordinate is unspecified. Whereas with your wording, a 1D coordinate should be interpreted as a horizontal coordinate. So gnome-shell will need to give 2D coordinates to communicate that its workspaces are vertical, but other compositors can use 1D coords for their horizontal workspaces.

Reply:
gnome-shell should do no such thing - clients should let the user specify how to arrange their workspaces, vertically or horizontally, in 1D. X/Y isn't meaningful in 1 dimension, so the X coordinate being arranged vertically isn't a problem. When transposing between coordinate systems you can losslessly transpose one coordinate onto "another" - the X coordinate in 2D doesn't mean the same thing as the X coordinate in 1D.

Reply:
What do you propose gnome-shell send to clients to tell them that if they want to be consistent with how gnome-shell represents the workspaces in its overview interface, that they should represent the workspaces vertically?

Even if this was set by a client, it needs to be communicated to subsequent clients. Sending 2D coords to communicate this seems to me like a legitimate and unproblematic way to do it.

Reply:
>What do you propose gnome-shell send to clients to tell them that if they want to be consistent with how gnome-shell represents the workspaces in its overview interface, that they should represent the workspaces vertically?

I would suggest that it doesn't. Besides - what if the user wants their workspaces shown horizontally in their panel and vertically in GNOME? We shouldn't add a dimension that doesn't exist.

Reply:
That's acceptable to me, but would be a bit strange because you could click on a workspace oriented to the right of the current one in your panel app, and gnome-shell would show a vertical animation. Windows overlapping multiple workspaces would display strangely if the panel app showed their positions with squares.

I'd argue the dimension exists because of these animations and window overlapping. But I'm happy to leave it unspecified and leave it up to gnome-shell to decide what to send - it can send a 2d coord if it wants.

Reply:
guys, I really think we should let it be in some free form like a string. There are many different combinations that are possible, and that one here doesn't address tagging wm's.

Reply:
>guys, I really think we should let it be in some free form like a string. There are many different combinations that are possible, and that one here doesn't address tagging wm's.

+1. I wouldn't address coords.

It should be enough to get the output layout using xdg-output, and use the workspace::enter,leave events to know non which output(s) a workspace is on.

Reply:
That would work for sway, but not for GNOME, for which workspaces are arranged at a higher level than outputs.

Reply:
as I said, we can just make it a string and say that compositors arranging in a grid would follow `x-y-z-...` format or whatever we find convenient. This way we also leave space for compositors that do not arrange in a grid, and they can then use their own specific way to encode data using that field in the protocol. IMO there is no one-size-fits-all solution.

Reply:
plus, in Wayfire I can't really make workspaces arranged in a grid in multimonitor configuration, they aren't in the same coordinate system.

Reply:
The name field works for tagging, no?

Coordinates seem common enough that if we leave it free-form, I imagine an ad-hoc standard for making strings of coordinates will present itself (and that gnome would end up sending 2D coordinates within the strings to tell clients that its workspaces are vertical). So by making it a string we're just deferring that standard to be a de-facto one instead of in the protocol. The benefit is allowing information about non-grid-like layouts.

But if it's an opaque string, what's it called? 'layout?' 'geometry'?

My preferred outcome is coordinates with strictly defined interpretation as x, y, z, etc, with compositors without such a concept just not using this part of the protocol. We can add extra events for other paradigms, including a catch-all opaque string for compositor-specific info.

Reply:
hm okay, if we make this event optional, I'm ok

Reply:
> plus, in Wayfire I can't really make workspaces arranged in a grid in multimonitor configuration, they aren't in the same coordinate system.

The coordinates are only meaningful among workspaces that share a set of outputs. So you can have a workspace at (0,0) on output 1, and a workspace on (0,0) on output 2. Or you can have a workspace at (0,0) that is attached to both outputs 1 and 2. But you cannot have *two* workspaces with coords (0,0) on the same outputs. This all may have been more clear with the original concept of 'workspace groups' - that the coordinates are only meaningful within a  workspace group that share a common set of outputs.

Reply:
>an ad-hoc standard for making strings of coordinates will present itself

This sounds *fucking awful*.

Reply:
> This sounds _fucking awful_.

:+1: